### DISCUSSION TITLE
Problem installing Octoeverywhere on Ender3v3

### DISCUSSION THREAD

**Discussion by dkahn93** [Guilouz/Creality-Helper-Script-Wiki#707] [Issues]
This is the full log:

                        OctoEverywhere                        │
 ├──────────────────────────────────────────────────────────────┤
 │                                                              │
 │ Cloud empower your Klipper printers with free, private, and  │
 │ unlimited remote access to your full web control portal from │
 │ anywhere!                                                    │
 │                                                              │
 └──────────────────────────────────────────────────────────────┘

 Are you sure you want to install OctoEverywhere ? (y/n): y

Info: Downloading OctoEverywhere...
Cloning into '/usr/data/octoeverywhere'...
remote: Enumerating objects: 2785, done.
remote: Counting objects: 100% (1837/1837), done.
remote: Compressing objects: 100% (574/574), done.
remote: Total 2785 (delta 1445), reused 1525 (delta 1248), pack-reused 948 (from 1)
Receiving objects: 100% (2785/2785), 1.20 MiB | 1.21 MiB/s, done.
Resolving deltas: 100% (1957/1957), done.
Info: Running OctoEverywhere installer...

      @@@@@@@@@@@@@@@@@@@@@@@@***@@@@@@@@@@@@@@@@@@@@@@@
      @@@@@@@@@@@@@@***********************@@@@@@@@@@@@@
      @@@@@@@@@@*******************************@@@@@@@@@
      @@@@@@@@***********************************@@@@@@@
      @@@@@,,,************************/////////*****@@@@
      @@@@,,,,,,*****************//////////////******@@@
      @@,,,,,,,,,,***********//////////////////*******@@
      @@,,,,,,,,,,,,*******////////****///////*********@
      @,,,,,,,,,,,/////////////////****//////***********
      @,,,,,,,//////////////////////////////************
      ,,,,,,,,////////////////////////////**************
      @,,,,,,,,,,,,/////////////////////****************
      @,,,,,,,,,,,,,,/////////////////******************
      @@,,,,,,,,,,,,,,,,//////////////*****************@
      @@@,,,,,/#######,,,,///////////*****************@@
      @@@@,,,##########,,,,,,,//////,****************@@@
      @@@@@,##########,,,,,,,,,////,,,,*************@@@@
      @@@@@########,,,,,,,,,,,,//,,,,,,,,********@@@@@@@
      @@@@@#@@@@,,,,,,,,,,,,,,,,,,,,,,,,,,,***,@@@@@@@@@
      @@@@@@@@@@@@@@@,,,,,,,,,,,,,,,,,,,,,@@@@@@@@@@@@@@

  OctoEverywhere For Klipper, Creality OS, And Bambu Lab Printers
 The 3D Printing Communities #1 Remote Access And AI Cloud Service

OctoEverywhere empowers the worldwide maker community with...
  - Free & Unlimited Mainsail, Fluidd, And Bambu Lab Printers Remote Access
  - Free & Unlimited Next-Gen AI Print Failure Detection
  - Free Full Frame Rate & Full Resolution Webcam Streaming
  - 5 Star Rated iOS & Android Apps
  - Real-Time Print Notifications
  - And So Much More

Running in K1 and K1 Max OS mode
Checking required system packages are installed...
Downloading http://bin.entware.net/mipselsf-k3.4/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/entware
Installing python3 (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3_3.11.7-1_mipsel-3.4.ipk
Installing libpython3 (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libpython3_3.11.7-1_mipsel-3.4.ipk
Installing python3-base (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-base_3.11.7-1_mipsel-3.4.ipk
Installing libbz2 (1.0.8-1a) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libbz2_1.0.8-1a_mipsel-3.4.ipk
Installing zlib (1.3.1-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/zlib_1.3.1-1_mipsel-3.4.ipk
Installing python3-light (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-light_3.11.7-1_mipsel-3.4.ipk
Installing python3-asyncio (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-asyncio_3.11.7-1_mipsel-3.4.ipk
Installing python3-email (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-email_3.11.7-1_mipsel-3.4.ipk
Installing python3-cgi (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-cgi_3.11.7-1_mipsel-3.4.ipk
Installing python3-pydoc (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-pydoc_3.11.7-1_mipsel-3.4.ipk
Installing python3-cgitb (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-cgitb_3.11.7-1_mipsel-3.4.ipk
Installing python3-codecs (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-codecs_3.11.7-1_mipsel-3.4.ipk
Installing libffi (3.4.6-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libffi_3.4.6-1_mipsel-3.4.ipk
Installing python3-ctypes (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-ctypes_3.11.7-1_mipsel-3.4.ipk
Installing libgdbm (1.23-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libgdbm_1.23-1_mipsel-3.4.ipk
Installing python3-dbm (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-dbm_3.11.7-1_mipsel-3.4.ipk
Installing python3-decimal (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-decimal_3.11.7-1_mipsel-3.4.ipk
Installing python3-distutils (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-distutils_3.11.7-1_mipsel-3.4.ipk
Installing python3-logging (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-logging_3.11.7-1_mipsel-3.4.ipk
Installing liblzma (5.4.6-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/liblzma_5.4.6-1_mipsel-3.4.ipk
Installing python3-lzma (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-lzma_3.11.7-1_mipsel-3.4.ipk
Installing python3-multiprocessing (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-multiprocessing_3.11.7-1_mipsel-3.4.ipk
Installing libncursesw (6.4-3) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libncursesw_6.4-3_mipsel-3.4.ipk
Installing python3-ncurses (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-ncurses_3.11.7-1_mipsel-3.4.ipk
Installing libatomic (8.4.0-11) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libatomic_8.4.0-11_mipsel-3.4.ipk
Installing libopenssl (3.0.14-2) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libopenssl_3.0.14-2_mipsel-3.4.ipk
Installing ca-certificates (20240203-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/ca-certificates_20240203-1_all.ipk
Installing python3-openssl (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-openssl_3.11.7-1_mipsel-3.4.ipk
Installing libreadline (8.2-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libreadline_8.2-1_mipsel-3.4.ipk
Installing python3-readline (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-readline_3.11.7-1_mipsel-3.4.ipk
Installing libsqlite3 (3410200-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libsqlite3_3410200-1_mipsel-3.4.ipk
Installing python3-sqlite3 (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-sqlite3_3.11.7-1_mipsel-3.4.ipk
Installing python3-unittest (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-unittest_3.11.7-1_mipsel-3.4.ipk
Installing python3-urllib (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-urllib_3.11.7-1_mipsel-3.4.ipk
Installing libuuid (2.40.1-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libuuid_2.40.1-1_mipsel-3.4.ipk
Installing python3-uuid (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-uuid_3.11.7-1_mipsel-3.4.ipk
Installing python3-xml (3.11.7-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-xml_3.11.7-1_mipsel-3.4.ipk
Installing python3-pip (23.3.1-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-pip_23.3.1-1_mipsel-3.4.ipk
Installing python3-pillow (10.1.0-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/python3-pillow_10.1.0-1_mipsel-3.4.ipk
Installing libpng (1.6.43-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libpng_1.6.43-1_mipsel-3.4.ipk
Installing libfreetype (2.13.2-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libfreetype_2.13.2-1_mipsel-3.4.ipk
Installing libjpeg-turbo (3.0.3-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libjpeg-turbo_3.0.3-1_mipsel-3.4.ipk
Installing libtiff (4.6.0-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libtiff_4.6.0-1_mipsel-3.4.ipk
Installing libwebp (1.4.0-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/libwebp_1.4.0-1_mipsel-3.4.ipk
Configuring libpython3.
Configuring python3-base.
Configuring libbz2.
Configuring zlib.
Configuring python3-light.
Configuring python3-email.
Configuring python3-urllib.
Configuring libatomic.
Configuring python3-pydoc.
Configuring liblzma.
Configuring python3-cgi.
Configuring python3-cgitb.
Configuring python3-decimal.
Configuring libuuid.
Configuring python3-uuid.
Configuring python3-xml.
Configuring libpng.
Configuring libfreetype.
Configuring libjpeg-turbo.
Configuring libtiff.
Configuring libwebp.
Configuring python3-asyncio.
Configuring python3-codecs.
Configuring libffi.
Configuring python3-ctypes.
Configuring libgdbm.
Configuring python3-dbm.
Configuring python3-distutils.
Configuring python3-logging.
Configuring python3-lzma.
Configuring python3-multiprocessing.
Configuring libncursesw.
Configuring python3-ncurses.
Configuring libopenssl.
Configuring ca-certificates.
Configuring python3-openssl.
Configuring libreadline.
Configuring python3-readline.
Configuring libsqlite3.
Configuring python3-sqlite3.
Configuring python3-unittest.
Configuring python3.
Configuring python3-pillow.
Configuring python3-pip.
Checking Python Virtual Environment For OctoEverywhere...
No virtual environment found, creating one now.
ERROR:root:code for hash md5 was not found.
Traceback (most recent call last):
  File "/usr/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type md5
ERROR:root:code for hash sha1 was not found.
Traceback (most recent call last):
  File "/usr/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha1
ERROR:root:code for hash sha224 was not found.
Traceback (most recent call last):
  File "/usr/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha224
ERROR:root:code for hash sha256 was not found.
Traceback (most recent call last):
  File "/usr/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha256
ERROR:root:code for hash sha384 was not found.
Traceback (most recent call last):
  File "/usr/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha384
ERROR:root:code for hash sha512 was not found.
Traceback (most recent call last):
  File "/usr/lib/python2.7/hashlib.py", line 147, in <module>
    globals()[__func_name] = __get_hash(__func_name)
  File "/usr/lib/python2.7/hashlib.py", line 97, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type sha512
Traceback (most recent call last):
  File "/usr/bin/virtualenv", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3250, in <module>
    @_call_aside
  File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3234, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3263, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 583, in _build_master
    ws.require(__requires__)
  File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 900, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python2.7/site-packages/pkg_resources/__init__.py", line 786, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'virtualenv==15.1.0' distribution was not found and is required by the application
Updating PIP if needed... (this can take a few seconds or so)
./install.sh: line 294: /usr/data/octoeverywhere-env/bin/python: not found
Installing or updating required python libs...
./install.sh: line 303: /usr/data/octoeverywhere-env/bin/pip3: not found
./install.sh: line 434: /usr/data/octoeverywhere-env/bin/python3: not found

 ✓ OctoEverywhere has been installed successfully!

**Comment by EchotreePrints** [comment]
I'm about to post a way to fix this, as well as for mobileraker issues.

The short version is that you'll need to ssh to the printer, and execute the following commands in order:

For octoeverywhere:
~~`sudo rm -rf /usr/data/octoeverywehre-env /usr/data/octoeverywhere`~~

`sudo rm -rf /usr/data/octoeverywhere-env /usr/data/octoeverywhere`
This will remove the octoeverywhere venv and directory that are installed with the helper script.

`cd /usr/data/`
moves our active directory to the /usr/data directory.

`git clone https://github.com/QuinnDamerell/OctoPrint-OctoEverywhere.git octoeverywhere`
clones the latest version of the octoeverywhere plugin to the usr/data/octoeverywhere dir, effectively replacing the version originally installed with the helper script

`cd octoeverywhere`
moves us into the octoeverywhere dir

~~`sh isntall.sh`~~
`sh install.sh`
runs the install script to reinstall the good version.

For whatever reason, when run through the helper script, octoeverywhere and sometimes mobileraker will fail to properly install their virtual environment, which prevents them from running, and from being accurately uninstalled by the helper-script. In order to make this work, you'll have to manually remove the affected venv and source folder and manually reinstall them/add them to the moonraker update manager configuration. 

[Edit] fixed the typo

**Comment by dkahn93** [comment]
It works! Thank you so much. One note you misspelled install.sh.

**Comment by dkahn93** [comment]
Today, I updated my firmware, did a factory reset, and then reinstalled creality helper. This fix doesn't seem to be working anymore.
This is what it shows:
`Checking required system packages are installed...
Downloading http://bin.entware.net/mipselsf-k3.4/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/entware
Package python3 (3.11.10-1) installed in root is up to date.
Package python3-pip (23.3.1-1) installed in root is up to date.
Package python3-pillow (10.1.0-1) installed in root is up to date.
Checking Python Virtual Environment For OctoEverywhere...
No virtual environment found, creating one now.
install.sh: line 208: /opt/bin/virtualenv: not found
Updating PIP if needed... (this can take a few seconds or so)
install.sh: line 294: /usr/data/octoeverywhere-env/bin/python: not found
Installing or updating required python libs...
install.sh: line 303: /usr/data/octoeverywhere-env/bin/pip3: not found
install.sh: line 434: /usr/data/octoeverywhere-env/bin/python3: not found`

**Reply by EchotreePrints** [reply]
Hey! Sorry about that, I fixed a couple of typos in my main paragraph. I am dyslexic so bear with me on this lol. 

The first line is supposed to remove the existing octoeverywhere-env directory, but if it's misspelled like in the original paragraph, it won't do anything.

Try again with the typos fixed. I haven't updated my firmware yet, I usually wait a while between updates for reasons like this.

**Reply by hmast37** [reply]
Solved already, just needed to reboot. I don't remember needing to do this the first time around. So far after restoring the firmware on the nebula pad there's been a few hickups. Maybe my unit is malfunctioning.

I installed the latest rooted firmware on the retail nebula pad myself now, 1.1.0.27, and even running the git-cloned version I am getting python errors.
No virtual environment found, creating one now.
./install.sh: line 219: /opt/bin/virtualenv: not found
Updating PIP if needed... (this can take a few seconds or so)
./install.sh: line 319: /usr/data/octoeverywhere-env/bin/python: not found
Installing or updating required python libs...
./install.sh: line 329: /usr/data/octoeverywhere-env/bin/pip3: not found
./install.sh: line 469: /usr/data/octoeverywhere-env/bin/python3: not found

Actually I think I just needed to reboot, it seems to be installing now.
